docs/backend/schemas.md

```diff
@@ -0,0 +1,622 @@
+Each value is atomically constrained by *value_type* schema field.
+
+Example:
+
+```
+from zope import schema
+from plone.supermodel import model
+from plone.autoform import directives as form
+
+from z3c.form.browser.checkbox import CheckBoxFieldWidget
+
+class IMultiChoice(model.Schema):
+    ...
+
+    # Contains lists of values from Choice list using special "get_field_list" vocabulary
+    # We also give a plone.autoform.directives hint to render this as
+    # multiple checbox choices
+    form.widget(yourField=CheckBoxFieldWidget)
+    yourField = schema.List(title=u"Available headers and animations",
+                               description=u"Headers and animations uploaded here",
+                               required=False,
+                               value_type=zope.schema.Choice(source=yourVocabularyFunction),
+                               )
+```
+
+### Dynamic schemas
+
+Schemas are singletons, as there only exist one class instance
+per Python run-time. For example, if you need to feed schemas generated dynamically
+to form engine, you need to
+
+- If the form engine (e.g. z3c.form refers to schema fields, then
+  replace these references with dynamically generated copes)
+- Generate a Python class dynamically. Output Python source code,
+  then `eval()` it. Using `eval()` is almost always considered
+  as a bad practice.
+
+```{warning}
+Though it is possible, you should not modify zope.schema classes
+in-place
+as the same copy is shared between different threads and
+if there are two concurrent HTTP requests problems occur.
+```
+
+#### Replacing schema fields with dynamically modified copies
+
+The below is an example for z3c.form. It uses Python `copy`
+module to copy f.field reference, which points to zope.schema
+field. For this field copy, we modify *required* attribute based
+on input.
+
+Example:
+
+```
+@property
+def fields(self):
+    """ Get the field definition for this form.
+
+    Form class's fields attribute does not have to
+    be fixed, it can be property also.
+    """
+
+    # Construct the Fields instance as we would
+    # normally do in more static way
+    fields = z3c.form.field.Fields(ICheckoutAddress)
+
+    # We need to override the actual required from the
+    # schema field which is a little tricky.
+    # Schema fields are shared between instances
+    # by default, so we need to create a copy of it
+    if self.optional:
+        for f in fields.values():
+            # Create copy of a schema field
+            # and force it unrequired
+            schema_field = copy.copy(f.field) # shallow copy of an instance
+            schema_field.required = False
+            f.field = schema_field
+```
+
+#### Don't use dict {} or list \[\] as a default value
+
+Because how Python object construction works, giving \[\] or {}
+as a default value will make all created field values to share this same object.
+
+<http://effbot.org/zone/default-values.htm>
```

Comment:
That site and the human behind effbot have passed away.

https://lwn.net/Articles/878325/

We could change the link to:

https://web.archive.org/web/20201112004749/https://www.effbot.org/zone/default-values.htm

...but I really don't want to rely on external outdated content, especially on web.archive.org because it goes down so often, and external sites tend to go away over time.

Reply:
What about
- https://docs.python-guide.org/writing/gotchas/
- https://docs.python.org/3/faq/programming.html#why-are-default-values-shared-between-objects

Reply:
i added the gotcha link ;)